### PR TITLE
Use Google Cloud API to do the image analysis.

### PR DESCRIPTION
**Pull request by jiaxi312**
I'm trying to use the google cloud api to analyse an image. But, it still gives me an access error. 
The screen shot is https://screenshot.googleplex.com/oiPCqDVMjUH.png

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -82,6 +85,18 @@ private static void showUploadedComment(PrintWriter out, String comment, String
     out.println("<a href=\"/index.html\"> click here to return </a>");
   }
 
+  /** Shows the labels extracted from the uploaded impage*/
+  private static void showImageLabels(HttpServletRequest request) throws IOException{
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: extra space after IOException?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -69,6 +71,7 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     ServletHelper.DEFAULT_DATASTORE_SERVICE.put(commentEntity);
 
     showUploadedComment(response.getWriter(), comment, imageUrl);
+    showImageLabels(request);
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:74]: Is it just for debuging?

File: portfolio/src/main/java/com/google/sps/servlets/ServletHelper.java [hunk 3]
```diff
@@ -75,6 +83,87 @@ protected static String getUploadedFileUrl(HttpServletRequest request, String fo
...
+   * Blobstore stores files as binary data. This function retrieves the binary data stored at the
+   * BlobKey parameter.
+   */
+  protected static byte[] getBlobBytes(BlobKey blobKey) throws IOException {
+    ByteArrayOutputStream outputBytes = new ByteArrayOutputStream();
+
+    int fetchSize = BlobstoreService.MAX_BLOB_FETCH_SIZE;
+    long currentByteIndex = 0;
+    boolean continueReading = true;
+    while (continueReading) {
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/ServletHelper.java]: optional: I would suggest while true, break, which is equivalent.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (jiaxi312) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Fixed

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (jiaxi312) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:74]: Not really. For now, the tutorial says just print the labels to console in order to make sure the code works. After this step, the labels will be displayed in the webpage.
But, still we don't have the access to use this library. So, access error will show in real execution.

File: portfolio/src/main/java/com/google/sps/servlets/ServletHelper.java (hunk 3 above)
Comment (jiaxi312) [reply portfolio/src/main/java/com/google/sps/servlets/ServletHelper.java]: Fixed

File: portfolio/src/main/java/com/google/sps/servlets/ServletHelper.java [hunk 4]
```diff
@@ -75,6 +83,86 @@ protected static String getUploadedFileUrl(HttpServletRequest request, String fo
...
+    // User submitted form without selecting a file, so we can't get a BlobKey. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so the BlobKey is empty. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/ServletHelper.java]: also make ```BlobInfoFactory``` private static final?

File: portfolio/src/main/java/com/google/sps/servlets/ServletHelper.java [hunk 5]
```diff
@@ -75,6 +83,86 @@ protected static String getUploadedFileUrl(HttpServletRequest request, String fo
...
+    Feature feature = Feature.newBuilder().setType(Feature.Type.LABEL_DETECTION).build();
+    AnnotateImageRequest request =
+        AnnotateImageRequest.newBuilder().addFeatures(feature).setImage(image).build();
+    List<AnnotateImageRequest> requests = new ArrayList<>();
+    requests.add(request);
+
+    ImageAnnotatorClient client = ImageAnnotatorClient.create();
+    BatchAnnotateImagesResponse batchResponse = client.batchAnnotateImages(requests);
+    client.close();
+    List<AnnotateImageResponse> imageResponses = batchResponse.getResponsesList();
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/ServletHelper.java]: I would suggest you use the API like the example code: https://googleapis.dev/java/google-cloud-vision/latest/com/google/cloud/vision/v1/ImageAnnotatorClient.html#batchAnnotateImages-java.util.List-

File: portfolio/src/main/java/com/google/sps/servlets/ServletHelper.java [hunk 6]
```diff
@@ -75,6 +83,86 @@ protected static String getUploadedFileUrl(HttpServletRequest request, String fo
...
+    Image image = Image.newBuilder().setContent(byteString).build();
+
+    Feature feature = Feature.newBuilder().setType(Feature.Type.LABEL_DETECTION).build();
+    AnnotateImageRequest request =
+        AnnotateImageRequest.newBuilder().addFeatures(feature).setImage(image).build();
+    List<AnnotateImageRequest> requests = new ArrayList<>();
+    requests.add(request);
+
+    ImageAnnotatorClient client = ImageAnnotatorClient.create();
+    BatchAnnotateImagesResponse batchResponse = client.batchAnnotateImages(requests);
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/ServletHelper.java]: will ```Arrays.asList()``` help you construct requests inline?

File: portfolio/src/main/java/com/google/sps/servlets/ServletHelper.java (hunk 4 above)
Comment (jiaxi312) [reply portfolio/src/main/java/com/google/sps/servlets/ServletHelper.java]: Done!

File: portfolio/src/main/java/com/google/sps/servlets/ServletHelper.java (hunk 5 above)
Comment (jiaxi312) [reply portfolio/src/main/java/com/google/sps/servlets/ServletHelper.java]: I imitated the example code and catch the ApiException if the remote call fails by some reason.

File: portfolio/src/main/java/com/google/sps/servlets/ServletHelper.java (hunk 6 above)
Comment (jiaxi312) [reply portfolio/src/main/java/com/google/sps/servlets/ServletHelper.java]: Done!

**Review by zixiyan — APPROVED** [review]

**Pull request merged**